web/modules/custom/nfafmis/nfafmis.module

```diff
@@ -798,3 +759,199 @@ function nfafmis_form_ajax_validate(array &$form, FormStateInterface $form_state
   $form_state->setResponse($response);
   return $response;
 }
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ */
+function nfafmis_form_user_form_alter(array &$form, FormStateInterface $form_state) {
+  // Show the Management unit/Range field for Range users only.
+  $form['management_unit']['#states'] = [
+    'visible' => [
+      [':input[name="roles[range_user]"]' => ['checked' => TRUE]],
+      [':input[name="roles[range_power_user]"]' => ['checked' => TRUE]],
+    ],
+  ];
+  $form['management_unit']['widget']['0']['target_id']['#states'] = [
+    'required' => [
+      [':input[name="roles[range_user]"]' => ['checked' => TRUE]],
+      [':input[name="roles[range_power_user]"]' => ['checked' => TRUE]],
+    ],
+  ];
+
+  // Only users with account admin permissions can edit the management unit.
+  if (!\Drupal::currentUser()->hasPermission('administer users')) {
+    $form['management_unit']['#disabled'] = 'disabled';
+  }
+}
+
+/**
+ * Implements hook_node_access_records().
+ * Set the rules that determine who has access to this $node.
+ */
+function nfafmis_node_access_records(NodeInterface $node) {
+  $grants = [];
+
+  // Content types that will be restricted according to the grants of the user
+  // accessing the node.
+  $bundles = ['farmer_details', 'offer_license', 'sub_area'];
+  if (!in_array($node->bundle(), $bundles) || !$node->isPublished()) {
+    // Access to other content types will be determined by standard permissions.
+    return $grants;
+  }
+
+  /** @var \Drupal\Core\Config\Entity\ConfigEntityStorage $entity_storage */
+  $node_storage = \Drupal::entityTypeManager()->getStorage('node');
+  $range = NULL;
+
+  if ($node->bundle() == 'farmer_details') {
+    // Retrieve the Areas that the farmer has allocations in, and from there the
+    // CFRs and Ranges (Management Unit) that the area is in.
+    // Load Sub area nodes ids.
+    $nids = \Drupal::entityQuery('node')
+      ->condition('type', 'offer_license')
+      ->condition('field_farmer_name_ref', $node->id())
+      ->accessCheck(FALSE)
+      ->execute();
+    $areas = $node_storage->loadMultiple($nids);
+
+    foreach ($areas as $area) {
+      $range = $area->field_central_forest_reserve->entity->management_unit->entity;
+      // Range Users are granted 'view' permission to nodes in their range.
+      // Remove duplicates by setting explicit array key.
+      $grants[$range->id()] = [
+        'realm' => nfafmis_range_access_realm($range),
+        'gid' => $range->id(),
+        'grant_view' => 1,
+        'grant_update' => 0,
+        'grant_delete' => 0,
+        'priority' => 0,
+      ];
+      // Range Power Users are granted all permissions to nodes in their range.
+      $grants[$range->id() + 10000] = [
```

Comment:
It's there to ensure that the grant is added as a new array item and doesn't overwrite the item added at line 818. At line 818 I add items to the array with `$grants[$range->id()]` instead of `$grants[]` because we can't have duplicate grants with the same realm and gid - duplicates cause a db integrity constraint violation. A farmer can have multiple areas in the same range and using  `$grants[$range->id()]` solves the integrity constraint violation. So now when adding the second power user grant I can't use the same $range->id() index. Can you think of a more elegant way to do this?

Reply:
Ah I see, feels a bit arbitrary, and at least grants a comment on why we do this.
I would query to see what's the last one, but Idk if it would be even possible at this stage.